File: src/movieshelf/cache.py
"""Local cache for matched metadata, artwork, and cleaned grid titles.

Lives under ``%APPDATA%\\MovieShelf\\cache`` (a sibling of ``logs``):
    cache/movies/<hash>/meta.json      matched metadata
    cache/movies/<hash>/poster.jpg     downloaded poster
    cache/movies/<hash>/backdrop.jpg   downloaded backdrop
    cache/titles.json                  path -> cleaned grid title (with mtime)
"""

import base64
import hashlib
import json
import logging
import os

from .config import cache_dir

log = logging.getLogger('movieshelf.cache')


def _key(path: str) -> str:
    norm = os.path.normcase(os.path.normpath(path))
    return hashlib.sha1(norm.encode('utf-8')).hexdigest()


def movie_dir(path: str):
    d = cache_dir() / 'movies' / _key(path)
    d.mkdir(parents=True, exist_ok=True)
    return d


# -- matched metadata ---------------------------------------------------------
def load(path: str):
    meta_file = cache_dir() / 'movies' / _key(path) / 'meta.json'
    if not meta_file.exists():
        return None
    try:
        return json.loads(meta_file.read_text(encoding='utf-8'))
    except (ValueError, OSError) as exc:
        log.warning('Bad cache meta for %s: %s', path, exc)
        return None


def save(path: str, meta: dict) -> None:
    (movie_dir(path) / 'meta.json').write_text(json.dumps(meta, ensure_ascii=False, indent=2), encoding='utf-8')


def save_image(path: str, name: str, data: bytes) -> str:
    """Persist image bytes (e.g. 'poster.jpg') and return the file name stored."""
    (movie_dir(path) / name).write_bytes(data)
    return name


def image_data_uri(path: str, name: str) -> str:
    """Return a base64 data URI for a cached image, or '' if missing.

    Data URIs avoid WebView2 file:// cross-directory access restrictions.
    """
    if not name:
        return ''
    f = cache_dir() / 'movies' / _key(path) / name
    if not f.exists():
        return ''
    try:
        mime = 'image/png' if name.lower().endswith('.png') else 'image/jpeg'
        return f'data:{mime};base64,' + base64.b64encode(f.read_bytes()).decode('ascii')
    except OSError:
        return ''


# -- grid titles --------------------------------------------------------------
_titles = None


def _titles_file():
    return cache_dir() / 'titles.json'


def _load_titles() -> dict:
    global _titles
    if _titles is None:
        f = _titles_file()
        if f.exists():
            try:
                _titles = json.loads(f.read_text(encoding='utf-8'))
            except (ValueError, OSError):
                _titles = {}
        else:
            _titles = {}
    return _titles


def _mtime(path: str):
    try:
        return os.path.getmtime(path)
    except OSError:
        return 0.0


def get_title(path: str):
    entry = _load_titles().get(os.path.normcase(path))
    if entry and entry.get('mtime') == _mtime(path):
        return entry.get('title')
    return None


def set_title(path: str, title: str) -> None:
    titles = _load_titles()
    titles[os.path.normcase(path)] = {'title': title, 'mtime': _mtime(path)}
    try:
        _titles_file().write_text(json.dumps(titles, ensure_ascii=False), encoding='utf-8')
    except OSError as exc:
        log.debug('Could not write titles cache: %s', exc)

File: src/main/cache.ts
// Local cache for matched metadata, artwork, and cleaned grid titles. Port of cache.py.
//
//   cache/movies/<sha1>/meta.json      matched metadata
//   cache/movies/<sha1>/poster.jpg     downloaded poster
//   cache/movies/<sha1>/backdrop.jpg   downloaded backdrop
//   cache/titles.json                  path -> cleaned grid title (with mtime)
//
// The sha1 key and titles-key normalization intentionally match the Python app byte-for-byte
// so an existing %APPDATA%\MovieShelf\cache carries over.

import { createHash } from 'crypto'
import { existsSync, mkdirSync, readFileSync, writeFileSync, statSync, rmSync } from 'fs'
import { join } from 'path'
import { cacheDir } from './config'
import { normcase, normcaseNormpath } from './paths'
import { log } from './logging'
import type { Metadata } from '../../shared/types'

function key(path: string): string {
  return createHash('sha1').update(normcaseNormpath(path), 'utf-8').digest('hex')
}

export function movieDir(path: string): string {
  const d = join(cacheDir(), 'movies', key(path))
  mkdirSync(d, { recursive: true })
  return d
}

// -- matched metadata ---------------------------------------------------------
export function load(path: string): Metadata | null {
  const metaFile = join(cacheDir(), 'movies', key(path), 'meta.json')
  if (!existsSync(metaFile)) return null
  try {
    return JSON.parse(readFileSync(metaFile, 'utf-8')) as Metadata
  } catch (exc) {
    log.warn('Bad cache meta for', path, String(exc))
    return null
  }
}

export function save(path: string, meta: Metadata): void {
  writeFileSync(join(movieDir(path), 'meta.json'), JSON.stringify(meta, null, 2), 'utf-8')
}

export function saveImage(path: string, name: string, data: Buffer): string {
  writeFileSync(join(movieDir(path), name), data)
  return name
}

/** Base64 data URI for a cached image, or '' if missing (dodges file:// restrictions). */
export function imageDataUri(path: string, name: string): string {
  if (!name) return ''
  const f = join(cacheDir(), 'movies', key(path), name)
  if (!existsSync(f)) return ''
  try {
    const mime = name.toLowerCase().endsWith('.png') ? 'image/png' : 'image/jpeg'
    return `data:${mime};base64,` + readFileSync(f).toString('base64')
  } catch {
    return ''
  }
}

// -- grid titles --------------------------------------------------------------
interface TitleEntry {
  title: string
  mtime: number
}
let titles: Record<string, TitleEntry> | null = null

function titlesFile(): string {
  return join(cacheDir(), 'titles.json')
}

function loadTitles(): Record<string, TitleEntry> {
  if (titles === null) {
    const f = titlesFile()
    let loaded: Record<string, TitleEntry> = {}
    if (existsSync(f)) {
      try {
        loaded = JSON.parse(readFileSync(f, 'utf-8'))
      } catch {
        loaded = {}
      }
    }
    titles = loaded
  }
  return titles
}

function mtime(path: string): number {
  try {
    return statSync(path).mtimeMs / 1000 // seconds, matching Python os.path.getmtime
  } catch {
    return 0
  }
}

export function getTitle(path: string): string | null {
  const entry = loadTitles()[normcase(path)]
  if (entry && entry.mtime === mtime(path)) return entry.title
  return null
}

export function setTitle(path: string, title: string): void {
  const t = loadTitles()
  t[normcase(path)] = { title, mtime: mtime(path) }
  try {
    writeFileSync(titlesFile(), JSON.stringify(t), 'utf-8')
  } catch (exc) {
    log.debug('Could not write titles cache:', String(exc))
  }
}

// -- clear --------------------------------------------------------------------
/** Wipe all cached metadata, artwork, and grid titles; the next lookups re-fetch. */
export function clearCache(): void {
  const dir = cacheDir()
  rmSync(dir, { recursive: true, force: true })
  mkdirSync(dir, { recursive: true })
  titles = null // drop the in-memory titles cache so it reloads empty
  log.info('Cache cleared')
}
